m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+    }
+
+    @Override
+    public CachedGauge allocateGauge(String name, Map<String, String> tags) {
+        Metric gauge = newMetric(name, tags, MetricType.GAUGE);
+
+        return new CachedMetric(gauge, calculateSize(gauge));
+    }
+
+    @Override
+    public CachedTimer allocateTimer(String name, Map<String, String> tags) {
+        Metric timer = newMetric(name, tags, MetricType.TIMER);
+
+        return new CachedMetric(timer, calculateSize(timer));
+    }
+
+    @Override
+    public CachedHistogram allocateHistogram(String name, Map<String, String> tags, Buckets buckets) {
+        List<CachedHistogramBucketImpl> cachedValueBuckets = new ArrayList<>();
+        List<CachedHistogramBucketImpl> cachedDurationBuckets = new ArrayList<>();
+
+        int bucketIdLen = String.valueOf(buckets.size()).length();
+        bucketIdLen = Math.max(bucketIdLen, MIN_METRIC_BUCKET_ID_TAG_LENGTH);
+
+        String bucketIdFmt = String.format("%%0%sd", bucketIdLen);
+
+        BucketPair[] bucketPairs = BucketPairImpl.bucketPairs(buckets);
+
+        for (int i = 0; i < bucketPairs.length; i++) {
+            Map<String, String> valueTags = new HashMap<>(tags);
+            Map<String, String> durationTags = new HashMap<>(tags);
+
+            String idTagValue = String.format(bucketIdFmt, i);
+
+            valueTags.put(bucketIdTagName, idTagValue);
+            valueTags.put(bucketTagName, String.format("%s-%s",
+                valueBucketString(bucketPairs[i].lowerBoundValue()),
+                valueBucketString(bucketPairs[i].upperBoundValue())));
+
+            cachedValueBuckets.add(new CachedHistogramBucketImpl(
+                bucketPairs[i].upperBoundValue(),
+                bucketPairs[i].upperBoundDuration(),
+                (CachedMetric) allocateCounter(name, valueTags)
+            ));
+
+            durationTags.put(bucketIdTagName, idTagValue);
+            durationTags.put(bucketTagName, String.format("%s-%s",
+                durationBucketString(bucketPairs[i].lowerBoundDuration()),
+                durationBucketString(bucketPairs[i].upperBoundDuration())));
+
+            cachedDurationBuckets.add(new CachedHistogramBucketImpl(
+                bucketPairs[i].upperBoundValue(),
+                bucketPairs[i].upperBoundDuration(),
+                (CachedMetric) allocateCounter(name, durationTags)
+            ));
+        }
+
+        return new CachedHistogramImpl(
+            name,
+            tags,
+            buckets,
+            cachedValueBuckets,
+            cachedDurationBuckets
+        );
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            // A null metric signifies to processors that a flush was called,
+            // which causes the metric queue to be cleared
+            metricQueue.put(resourcePool.getSizedMetric());
+        } catch (InterruptedException e) {
```

Comment:
What circumstances would this fail under?

Reply:
The queue will block on the `put` call if the queue is already at capacity. If at that point, the thread is interrupted, we will reach this block.

Reply:
I don't know if we should ever be throwing an exception in steady state. I'd expect we at most log, never throw an exception once the scope is constructed. 

Could you confirm what the current Java client, or tally-go client do when they run into issues?